File: 133_clone_graph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if (node == null) return node;
        Map<Integer, UndirectedGraphNode> visited = new HashMap<Integer, UndirectedGraphNode>();
        UndirectedGraphNode result = dfs(node, visited);
        return result;
    }

    private UndirectedGraphNode dfs(UndirectedGraphNode original, Map<Integer, UndirectedGraphNode> visited){
        if (visited.containsKey(original.label)){
            return visited.get(original.label);
        }

        UndirectedGraphNode newNode = new UndirectedGraphNode(original.label); // 創建新的node
        visited.put(newNode.label, newNode);

        List<UndirectedGraphNode> neighbors = original.neighbors;
        for (UndirectedGraphNode neighbor: neighbors){ // 尋遍每個node
            UndirectedGraphNode newVisited = dfs(neighbor, visited);     // 繼續走訪新的neighbor
            newNode.neighbors.add(newVisited); // 把遇到的neighbor加到現有的neighbor
        }
        return newNode;
    }
}

File: python/133_clone_graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        visited = {}

        def dfs(node: 'Node'):
            if not node:
                return None
            if (node in visited):
                return visited[node]
            new_node = Node(node.val)
            visited[node] = new_node
            for nei in node.neighbors:
                new_node.neighbors.append(dfs(nei))
            return new_node

        res = dfs(node)
        return res


